lib/kube/grpc/grpc.go

```diff
@@ -0,0 +1,313 @@
+func (s *Server) ListKubernetesResources(ctx context.Context, req *proto.ListKubernetesResourcesRequest) (*proto.ListKubernetesResourcesResponse, error) {
+	userContext, err := s.authorize(ctx)
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+
+	if req.UseSearchAsRoles || req.UsePreviewAsRoles {
+		var extraRoles []string
+		if req.UseSearchAsRoles {
+			extraRoles = append(extraRoles, userContext.Checker.GetAllowedSearchAsRoles()...)
+		}
+		if req.UsePreviewAsRoles {
+			extraRoles = append(extraRoles, userContext.Checker.GetAllowedPreviewAsRoles()...)
+		}
+		clusterName, err := s.cfg.AuthClient.GetClusterName()
+		if err != nil {
+			return nil, trail.ToGRPC(err)
+		}
+		if err := userContext.UseExtraRoles(s.cfg.AuthClient, clusterName.GetClusterName(), extraRoles); err != nil {
+			return nil, trail.ToGRPC(err)
+		}
+		if err := s.emitAuditEvent(ctx, userContext, req); err != nil {
+			return nil, trail.ToGRPC(err)
+		}
+	}
+	// We use the unmapped identity here because Kube Proxy will handle
+	// the forwarding of the request to the correct leaf cluster if that's the case
+	// and it handles the mapping of the identity to the leaf cluster.
+	identity := userContext.UnmappedIdentity.GetIdentity()
+	identity.KubernetesCluster = req.KubernetesCluster
+	identity.Groups = userContext.Checker.RoleNames()
+	identity.RouteToCluster = req.TeleportCluster
+	switch {
+	case req.ResourceType == types.KindKubePod:
+		rsp, err := s.listKubernetesPods(ctx, identity, true, req)
+		return rsp, trail.ToGRPC(err)
+	default:
+		return nil, trail.ToGRPC(trace.BadParameter("unsupported resource type %q", req.ResourceType))
+	}
+}
+
+// authorize checks if the user is authorized to connect to the cluster.
+func (s *Server) authorize(ctx context.Context) (*auth.Context, error) {
+	accessDeniedMsg := "access denied"
+	userContext, err := s.cfg.Authz.Authorize(ctx)
+
+	switch {
+	case err == nil:
+		return userContext, nil
+	// propagate connection problem error so we can differentiate
+	// between connection failed and access denied
+	case trace.IsConnectionProblem(err):
+		return nil, trace.ConnectionProblem(err, "[07] failed to connect to the database")
+	case trace.IsAccessDenied(err):
+		// don't print stack trace, just log the warning
+		s.cfg.Log.Warn(err)
+		return nil, trace.AccessDenied(accessDeniedMsg)
+	case keys.IsPrivateKeyPolicyError(err):
+		// private key policy errors should be returned to the client
+		// unaltered so that they know to reauthenticate with a valid key.
+		return nil, trace.Unwrap(err)
+	default:
+		s.cfg.Log.Warn(trace.DebugReport(err))
+		return nil, trace.AccessDenied(accessDeniedMsg)
+	}
+}
+
+// emitAuditEvent emits an audit event for a resource search action and logs
+// the roles used to perform the search.
+func (s *Server) emitAuditEvent(ctx context.Context, userContext *auth.Context, req *proto.ListKubernetesResourcesRequest) error {
+	err := s.cfg.Emitter.EmitAuditEvent(
+		s.cfg.Context,
```

Comment:
I usually prefer to use the process context instead of `ctx` because if the context is canceled we can potentially miss the audit event but this case doesn't seem to be a problem when we miss them